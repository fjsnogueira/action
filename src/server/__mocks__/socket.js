const socket = {
  emit: jest.fn(),
  setAuthToken: jest.fn()
};

export default socket;
