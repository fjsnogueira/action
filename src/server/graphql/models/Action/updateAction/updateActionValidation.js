import legitify from 'universal/validation/legitify';

export default function updateActionValidation(tms) {
  return legitify({
    id: (value) => value
      .test((id) => {
        const [teamId] = id.split('::');
        return !tms.includes(teamId) && `Not on team ${teamId}`;
      }),
    // agendaId: (value) => value
    //   .test((id) => {
    //     if (!id) return undefined;
    //     const [teamId] = id.split('::');
    //     return !tms.includes(teamId) && `Not on team ${teamId}`;
    //   }),
    content: (value) => value
      .trim()
      .max(255, 'Whoa! That looks like 2 actions'),
    isComplete: (value) => value.boolean(),
    sortOrder: (value) => value.float(),
    // required for actions created in meetings
    teamMemberId: (value) => value
      .test((id) => {
        if (!id) return undefined;
        const [, teamId] = id.split('::');
        return !tms.includes(teamId) && `Not on team ${teamId}`;
      })
  });
}
