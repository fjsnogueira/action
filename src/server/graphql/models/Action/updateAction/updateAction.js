import getRethink from 'server/database/rethinkDriver';
import {ActionInput} from 'server/graphql/models/Action/actionSchema';
import {
  GraphQLNonNull,
  GraphQLBoolean,
} from 'graphql';
import {requireSUOrTeamMember, requireWebsocket} from 'server/utils/authorization';
import {handleSchemaErrors} from 'server/utils/utils';
import updateActionValidation from 'server/graphql/models/Action/updateAction/updateActionValidation';

export default {
  type: GraphQLBoolean,
  description: 'Update a action with a change in content, ownership, or status',
  args: {
    updatedAction: {
      type: new GraphQLNonNull(ActionInput),
      description: 'the updated action including the id, and at least one other field'
    },
  },
  async resolve(source, {updatedAction}, {authToken, socket}) {
    const r = getRethink();
    const now = new Date();

    // AUTH
    requireWebsocket(socket);
    const [teamId] = updatedAction.id.split('::');
    requireSUOrTeamMember(authToken, teamId);

    // VALIDATION
    const schema = updateActionValidation(authToken.tms);
    const {errors, data: {id: actionId, ...validAction}} = schema(updatedAction);
    handleSchemaErrors(errors);

    // RESOLUTION
    const newAction = {
      ...validAction,
      // don't update updatedAt for sortOrder changes
      updatedAt: (Object.keys(validAction).length > 1 || validAction.sortOrder === undefined) ? now : undefined
    };
    await r.table('Action').get(actionId).update(newAction);
    return true;
  }
};
