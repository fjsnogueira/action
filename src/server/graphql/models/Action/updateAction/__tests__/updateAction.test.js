import getRethink from 'server/database/rethinkDriver';
import mockAuthToken from 'server/__tests__/setup/mockAuthToken';
import MockDate from 'mockdate';
import {__now} from 'server/__tests__/setup/mockTimes';
import fetchAndSerialize from 'server/__tests__/utils/fetchAndSerialize';
import DynamicSerializer from 'dynamic-serializer';
import MockDB from 'server/__tests__/setup/MockDB';
import expectAsyncToThrow from 'server/__tests__/utils/expectAsyncToThrow';
import socket from 'server/__mocks__/socket';
import updateAction from 'server/graphql/models/Action/updateAction/updateAction';

MockDate.set(__now);
console.error = jest.fn();

describe('updateAction', () => {
  test('updates the sortOrder without changing updatedAt', async () => {
    // SETUP
    const r = getRethink();
    const dynamicSerializer = new DynamicSerializer();
    const mockDB = new MockDB();
    const {action, user} = await mockDB.init()
      .newAction();
    const actionId = action[0].id;
    const authToken = mockAuthToken(user[7]);

    // TEST
    const updatedAction = {
      id: actionId,
      sortOrder: 2
    };
    await updateAction.resolve(undefined, {updatedAction}, {authToken, socket});

    // VERIFY
    const db = await fetchAndSerialize({
      action: r.table('Action').get(actionId),
    }, dynamicSerializer);
    expect(db).toMatchSnapshot();
  });

  test('updates the content of the action', async () => {
    // SETUP
    const r = getRethink();
    const dynamicSerializer = new DynamicSerializer();
    const mockDB = new MockDB();
    const {action, user} = await mockDB.init()
      .newAction();
    const actionId = action[0].id;
    const authToken = mockAuthToken(user[7]);

    // TEST
    const updatedAction = {
      id: actionId,
      content: 'Updated content'
    };
    await updateAction.resolve(undefined, {updatedAction}, {authToken, socket});

    // VERIFY
    const db = await fetchAndSerialize({
      action: r.table('Action').get(actionId),
    }, dynamicSerializer);
    expect(db).toMatchSnapshot();
  });

  test('updates the status of the action', async () => {
    // SETUP
    const r = getRethink();
    const dynamicSerializer = new DynamicSerializer();
    const mockDB = new MockDB();
    const {action, user} = await mockDB.init()
      .newAction();
    const actionId = action[0].id;
    const authToken = mockAuthToken(user[7]);

    // TEST
    const updatedAction = {
      id: actionId,
      isComplete: true
    };
    await updateAction.resolve(undefined, {updatedAction}, {authToken, socket});

    // VERIFY
    const db = await fetchAndSerialize({
      action: r.table('Action').get(actionId),
    }, dynamicSerializer);
    expect(db).toMatchSnapshot();
  });

  test('updates the teamMemberId of the action (in meeting only)', async () => {
    // SETUP
    const r = getRethink();
    const dynamicSerializer = new DynamicSerializer();
    const mockDB = new MockDB();
    const {action, teamMember, user} = await mockDB.init()
      .newAction();
    const actionId = action[0].id;
    const authToken = mockAuthToken(user[7]);

    // TEST
    const updatedAction = {
      id: actionId,
      teamMemberId: teamMember[5].id
    };
    await updateAction.resolve(undefined, {updatedAction}, {authToken, socket});

    // VERIFY
    const db = await fetchAndSerialize({
      action: r.table('Action').get(actionId),
    }, dynamicSerializer);
    expect(db).toMatchSnapshot();
  });

  test('throws when no websocket is present', async () => {
    const authToken = {};
    await expectAsyncToThrow(updateAction.resolve(undefined, {updatedAction: {}}, {authToken}));
  });

  test('throw when the caller is not a team member', async () => {
    // SETUP
    const mockDB = new MockDB();
    const {action, user} = await mockDB.init()
      .newAction();
    const authToken = mockAuthToken(user[1], {tms: ['foo']});
    const actionId = action[0].id;

    // TEST
    const updatedAction = {
      id: actionId,
      isComplete: true
    };
    await expectAsyncToThrow(updateAction.resolve(undefined, {updatedAction}, {authToken, socket}), [mockDB.context.team.id]);
  });
});
