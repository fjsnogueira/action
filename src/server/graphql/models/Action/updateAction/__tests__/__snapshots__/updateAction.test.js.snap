// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateAction throw when the caller is not a team member 1`] = `"{\\"_error\\":\\"You do not have access to team 0\\"}"`;

exports[`updateAction throws when no websocket is present 1`] = `"{\\"_error\\":\\"this must be called from a websocket\\"}"`;

exports[`updateAction updates the content of the action 1`] = `
Object {
  "action": Object {
    "content": "Updated content",
    "createdAt": 2017-01-01T06:59:59.999Z,
    "createdBy": 1,
    "id": 0,
    "isComplete": false,
    "sortOrder": 0,
    "teamMemberId": 2,
    "updatedAt": 2017-01-01T08:00:00.000Z,
    "userId": 1,
  },
}
`;

exports[`updateAction updates the sortOrder without changing updatedAt 1`] = `
Object {
  "action": Object {
    "content": "Test Action[0]",
    "createdAt": 2017-01-01T06:59:59.999Z,
    "createdBy": 1,
    "id": 0,
    "isComplete": false,
    "sortOrder": 2,
    "teamMemberId": 2,
    "updatedAt": 2017-01-01T07:00:00.000Z,
    "userId": 1,
  },
}
`;

exports[`updateAction updates the status of the action 1`] = `
Object {
  "action": Object {
    "content": "Test Action[0]",
    "createdAt": 2017-01-01T06:59:59.999Z,
    "createdBy": 1,
    "id": 0,
    "isComplete": true,
    "sortOrder": 0,
    "teamMemberId": 2,
    "updatedAt": 2017-01-01T08:00:00.000Z,
    "userId": 1,
  },
}
`;

exports[`updateAction updates the teamMemberId of the action (in meeting only) 1`] = `
Object {
  "action": Object {
    "content": "Test Action[0]",
    "createdAt": 2017-01-01T06:59:59.999Z,
    "createdBy": 1,
    "id": 0,
    "isComplete": false,
    "sortOrder": 0,
    "teamMemberId": 2,
    "updatedAt": 2017-01-01T08:00:00.000Z,
    "userId": 1,
  },
}
`;
